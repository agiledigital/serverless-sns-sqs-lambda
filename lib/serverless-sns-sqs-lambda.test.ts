"use strict";

import CLI from "serverless/lib/classes/CLI";
import AwsProvider from "serverless/lib/plugins/aws/provider";
import Serverless from "serverless/lib/Serverless";
import ServerlessSnsSqsLambda from "./serverless-sns-sqs-lambda";

const slsOpt = {
  stage: "dev-test",
  region: "ap-southeast-2"
};

describe("Test Serverless SNS SQS Lambda", () => {
  let serverless;
  let serverlessSnsSqsLambda;
  beforeEach(() => {
    serverless = new Serverless();
    const options = {
      ...slsOpt
    };
    serverless.setProvider("aws", new AwsProvider(serverless));
    serverless.cli = new CLI(serverless);
    serverlessSnsSqsLambda = new ServerlessSnsSqsLambda(serverless, options);
  });

  afterEach(() => {
    jest.resetModules(); // reset modules after each test
  });

  it("should have one hook", () => {
    () => {
      expect(serverlessSnsSqsLambda.hooks.length).toBe(1);
      expect(serverlessSnsSqsLambda.hooks[0].keys).toBe(
        "aws:package:finalize:mergeCustomProviderResources"
      );
    };
  });

  it("should set the provider variable to an instance of AwsProvider", () =>
    expect(serverlessSnsSqsLambda.provider).toBeInstanceOf(AwsProvider));

  it("should have access to the serverless instance", () =>
    expect(serverlessSnsSqsLambda.serverless).toEqual(serverless));

  it("should set the options variable", () =>
    expect(serverlessSnsSqsLambda.options).toEqual({
      ...slsOpt
    }));

  it("should fail if name is not passed", () => {
    expect.assertions(1);
    expect(() => {
      serverlessSnsSqsLambda.validateConfig("func-name", "stage", {
        topicArn: "topicArn",
        name: undefined
      });
    }).toThrow(/name was \[undefined\]/);
  });

  it("should fail if topicArn is not passed", () => {
    expect.assertions(1);
    expect(() => {
      serverlessSnsSqsLambda.validateConfig("func-name", "stage", {
        topicArn: undefined,
        name: "name"
      });
    }).toThrow(/topicArn was \[undefined\]/);
  });

  describe("when no optional parameters are provided", () => {
    it("should produce valid SQS CF template items", () => {
      const template = { Resources: {} };
      const testConfig = {
        name: "some-name",
        topicArn: "arn:aws:sns:us-east-2:123456789012:MyTopic"
      };
      const validatedConfig = serverlessSnsSqsLambda.validateConfig(
        "test-function",
        "test-stage",
        testConfig
      );
      serverlessSnsSqsLambda.addEventQueue(template, validatedConfig);
      serverlessSnsSqsLambda.addEventDeadLetterQueue(template, validatedConfig);
      serverlessSnsSqsLambda.addEventSourceMapping(template, validatedConfig);
      serverlessSnsSqsLambda.addTopicSubscription(template, validatedConfig);
      expect(template).toMatchSnapshot();
    });
  });

  describe("when all parameters are provided", () => {
    it("should produce valid SQS CF template items", () => {
      const template = { Resources: {} };
      const testConfig = {
        name: "some-name",
        topicArn: "arn:aws:sns:us-east-2:123456789012:MyTopic",
        batchSize: 7,
        maximumBatchingWindowInSeconds: 99,
        prefix: "some prefix",
        maxRetryCount: 4,
        kmsMasterKeyId: "some key",
        kmsDataKeyReusePeriodSeconds: 200,
        deadLetterMessageRetentionPeriodSeconds: 1209600,
        enabled: false,
        visibilityTimeout: 999,
        rawMessageDelivery: true,
        filterPolicy: { pet: ["dog", "cat"] }
      };
      const validatedConfig = serverlessSnsSqsLambda.validateConfig(
        "test-function",
        "test-stage",
        testConfig
      );
      serverlessSnsSqsLambda.addEventQueue(template, validatedConfig);
      serverlessSnsSqsLambda.addEventDeadLetterQueue(template, validatedConfig);
      serverlessSnsSqsLambda.addEventSourceMapping(template, validatedConfig);
      serverlessSnsSqsLambda.addTopicSubscription(template, validatedConfig);
      expect(template).toMatchSnapshot();
    });
    it("should produce valid SQS FIFO CF template items", () => {
      const template = { Resources: {} };
      const testConfig = {
        name: "some-name",
        topicArn: "arn:aws:sns:us-east-2:123456789012:MyTopic",
        batchSize: 7,
        maximumBatchingWindowInSeconds: 99,
        prefix: "some prefix",
        maxRetryCount: 4,
        kmsMasterKeyId: "some key",
        kmsDataKeyReusePeriodSeconds: 200,
        fifoQueue: true,
        fifoThroughputLimit: "perMessageGroupId",
        deduplicationScope: "messageGroup",
        deadLetterMessageRetentionPeriodSeconds: 1209600,
        enabled: false,
        visibilityTimeout: 999,
        rawMessageDelivery: true,
        filterPolicy: { pet: ["dog", "cat"] }
      };
      const validatedConfig = serverlessSnsSqsLambda.validateConfig(
        "test-function",
        "test-stage",
        testConfig
      );
      serverlessSnsSqsLambda.addEventQueue(template, validatedConfig);
      serverlessSnsSqsLambda.addEventDeadLetterQueue(template, validatedConfig);
      serverlessSnsSqsLambda.addEventSourceMapping(template, validatedConfig);
      serverlessSnsSqsLambda.addTopicSubscription(template, validatedConfig);
      expect(template).toMatchSnapshot();
    });
  });

  describe("when encryption parameters are not provided", () => {
    it("should produce valid SQS CF template items", () => {
      const template = { Resources: {} };
      const testConfig = {
        name: "some-name",
        topicArn: "arn:aws:sns:us-east-2:123456789012:MyTopic",
        prefix: "some prefix",
        maxRetryCount: 4
      };
      const validatedConfig = serverlessSnsSqsLambda.validateConfig(
        "test-function",
        "test-stage",
        testConfig
      );
      serverlessSnsSqsLambda.addEventQueue(template, validatedConfig);
      serverlessSnsSqsLambda.addEventDeadLetterQueue(template, validatedConfig);
      serverlessSnsSqsLambda.addEventSourceMapping(template, validatedConfig);
      serverlessSnsSqsLambda.addTopicSubscription(template, validatedConfig);
      expect(template).toMatchSnapshot();
    });
  });

  describe("when overriding the generated CloudFormation template", () => {
    it("the overrides should take precedence", () => {
      const template = { Resources: {} };
      const testConfig = {
        name: "some-name",
        topicArn: "arn:aws:sns:us-east-2:123456789012:MyTopic",
        prefix: "some prefix",
        maxRetryCount: 4,
        enabled: true,
        visibilityTimeout: 1234,
        deadLetterMessageRetentionPeriodSeconds: 120,
        rawMessageDelivery: true,
        mainQueueOverride: {
          visibilityTimeout: 4321
        },
        deadLetterQueueOverride: {
          MessageRetentionPeriod: 1000
        },
        eventSourceMappingOverride: {
          Enabled: false
        },
        subscriptionOverride: {
          rawMessageDelivery: false
        }
      };
      const validatedConfig = serverlessSnsSqsLambda.validateConfig(
        "test-function",
        "test-stage",
        testConfig
      );
      serverlessSnsSqsLambda.addEventQueue(template, validatedConfig);
      serverlessSnsSqsLambda.addEventDeadLetterQueue(template, validatedConfig);
      serverlessSnsSqsLambda.addEventSourceMapping(template, validatedConfig);
      serverlessSnsSqsLambda.addTopicSubscription(template, validatedConfig);

      expect(template).toMatchSnapshot();
    });
  });
});
