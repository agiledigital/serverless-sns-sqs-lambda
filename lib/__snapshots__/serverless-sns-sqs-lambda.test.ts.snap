// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Serverless SNS SQS Lambda when all parameters are provided should produce valid SQS CF template items 1`] = `
Object {
  "Resources": Object {
    "Subscribesome-nameTopic": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "some-nameQueue",
            "Arn",
          ],
        },
        "FilterPolicy": Object {
          "pet": Array [
            "dog",
            "cat",
          ],
        },
        "Protocol": "sqs",
        "RawMessageDelivery": true,
        "TopicArn": "arn:aws:sns:us-east-2:123456789012:MyTopic",
      },
      "Type": "AWS::SNS::Subscription",
    },
    "Test-functionEventSourceMappingSQSsome-nameQueue": Object {
      "DependsOn": "IamRoleLambdaExecution",
      "Properties": Object {
        "BatchSize": 7,
        "Enabled": "False",
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "some-nameQueue",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "Test-functionLambdaFunction",
            "Arn",
          ],
        },
        "MaximumBatchingWindowInSeconds": 99,
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "some-nameDeadLetterQueue": Object {
      "Properties": Object {
        "KmsDataKeyReusePeriodSeconds": 200,
        "KmsMasterKeyId": "some key",
        "MessageRetentionPeriod": 1209600,
        "QueueName": "some prefixsome-nameDeadLetterQueue",
      },
      "Type": "AWS::SQS::Queue",
    },
    "some-nameQueue": Object {
      "Properties": Object {
        "KmsDataKeyReusePeriodSeconds": 200,
        "KmsMasterKeyId": "some key",
        "QueueName": "some prefixsome-nameQueue",
        "RedrivePolicy": Object {
          "deadLetterTargetArn": Object {
            "Fn::GetAtt": Array [
              "some-nameDeadLetterQueue",
              "Arn",
            ],
          },
          "maxReceiveCount": 4,
        },
        "VisibilityTimeout": 999,
      },
      "Type": "AWS::SQS::Queue",
    },
  },
}
`;

exports[`Test Serverless SNS SQS Lambda when all parameters are provided should produce valid SQS FIFO CF template items 1`] = `
Object {
  "Resources": Object {
    "Subscribesome-nameTopic": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "some-nameQueue",
            "Arn",
          ],
        },
        "FilterPolicy": Object {
          "pet": Array [
            "dog",
            "cat",
          ],
        },
        "Protocol": "sqs",
        "RawMessageDelivery": true,
        "TopicArn": "arn:aws:sns:us-east-2:123456789012:MyTopic",
      },
      "Type": "AWS::SNS::Subscription",
    },
    "Test-functionEventSourceMappingSQSsome-nameQueue": Object {
      "DependsOn": "IamRoleLambdaExecution",
      "Properties": Object {
        "BatchSize": 7,
        "Enabled": "False",
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "some-nameQueue",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "Test-functionLambdaFunction",
            "Arn",
          ],
        },
        "MaximumBatchingWindowInSeconds": 99,
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "some-nameDeadLetterQueue": Object {
      "Properties": Object {
        "DeduplicationScope": "messageGroup",
        "FifoQueue": true,
        "FifoThroughputLimit": "perMessageGroupId",
        "KmsDataKeyReusePeriodSeconds": 200,
        "KmsMasterKeyId": "some key",
        "MessageRetentionPeriod": 1209600,
        "QueueName": "some prefixsome-nameDeadLetterQueue",
      },
      "Type": "AWS::SQS::Queue",
    },
    "some-nameQueue": Object {
      "Properties": Object {
        "DeduplicationScope": "messageGroup",
        "FifoQueue": true,
        "FifoThroughputLimit": "perMessageGroupId",
        "KmsDataKeyReusePeriodSeconds": 200,
        "KmsMasterKeyId": "some key",
        "QueueName": "some prefixsome-nameQueue",
        "RedrivePolicy": Object {
          "deadLetterTargetArn": Object {
            "Fn::GetAtt": Array [
              "some-nameDeadLetterQueue",
              "Arn",
            ],
          },
          "maxReceiveCount": 4,
        },
        "VisibilityTimeout": 999,
      },
      "Type": "AWS::SQS::Queue",
    },
  },
}
`;

exports[`Test Serverless SNS SQS Lambda when encryption parameters are not provided should produce valid SQS CF template items 1`] = `
Object {
  "Resources": Object {
    "Subscribesome-nameTopic": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "some-nameQueue",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "RawMessageDelivery": false,
        "TopicArn": "arn:aws:sns:us-east-2:123456789012:MyTopic",
      },
      "Type": "AWS::SNS::Subscription",
    },
    "Test-functionEventSourceMappingSQSsome-nameQueue": Object {
      "DependsOn": "IamRoleLambdaExecution",
      "Properties": Object {
        "BatchSize": 10,
        "Enabled": "True",
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "some-nameQueue",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "Test-functionLambdaFunction",
            "Arn",
          ],
        },
        "MaximumBatchingWindowInSeconds": 0,
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "some-nameDeadLetterQueue": Object {
      "Properties": Object {
        "QueueName": "some prefixsome-nameDeadLetterQueue",
      },
      "Type": "AWS::SQS::Queue",
    },
    "some-nameQueue": Object {
      "Properties": Object {
        "QueueName": "some prefixsome-nameQueue",
        "RedrivePolicy": Object {
          "deadLetterTargetArn": Object {
            "Fn::GetAtt": Array [
              "some-nameDeadLetterQueue",
              "Arn",
            ],
          },
          "maxReceiveCount": 4,
        },
      },
      "Type": "AWS::SQS::Queue",
    },
  },
}
`;

exports[`Test Serverless SNS SQS Lambda when no optional parameters are provided should produce valid SQS CF template items 1`] = `
Object {
  "Resources": Object {
    "Subscribesome-nameTopic": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "some-nameQueue",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "RawMessageDelivery": false,
        "TopicArn": "arn:aws:sns:us-east-2:123456789012:MyTopic",
      },
      "Type": "AWS::SNS::Subscription",
    },
    "Test-functionEventSourceMappingSQSsome-nameQueue": Object {
      "DependsOn": "IamRoleLambdaExecution",
      "Properties": Object {
        "BatchSize": 10,
        "Enabled": "True",
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "some-nameQueue",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "Test-functionLambdaFunction",
            "Arn",
          ],
        },
        "MaximumBatchingWindowInSeconds": 0,
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "some-nameDeadLetterQueue": Object {
      "Properties": Object {
        "QueueName": "null-test-stage-Test-functionsome-nameDeadLetterQueue",
      },
      "Type": "AWS::SQS::Queue",
    },
    "some-nameQueue": Object {
      "Properties": Object {
        "QueueName": "null-test-stage-Test-functionsome-nameQueue",
        "RedrivePolicy": Object {
          "deadLetterTargetArn": Object {
            "Fn::GetAtt": Array [
              "some-nameDeadLetterQueue",
              "Arn",
            ],
          },
          "maxReceiveCount": 5,
        },
      },
      "Type": "AWS::SQS::Queue",
    },
  },
}
`;

exports[`Test Serverless SNS SQS Lambda when overriding the generated CloudFormation template the overrides should take precedence 1`] = `
Object {
  "Resources": Object {
    "Subscribesome-nameTopic": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "some-nameQueue",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "RawMessageDelivery": false,
        "TopicArn": "arn:aws:sns:us-east-2:123456789012:MyTopic",
      },
      "Type": "AWS::SNS::Subscription",
    },
    "Test-functionEventSourceMappingSQSsome-nameQueue": Object {
      "DependsOn": "IamRoleLambdaExecution",
      "Properties": Object {
        "BatchSize": 10,
        "Enabled": false,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "some-nameQueue",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "Test-functionLambdaFunction",
            "Arn",
          ],
        },
        "MaximumBatchingWindowInSeconds": 0,
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "some-nameDeadLetterQueue": Object {
      "Properties": Object {
        "MessageRetentionPeriod": 1000,
        "QueueName": "some prefixsome-nameDeadLetterQueue",
      },
      "Type": "AWS::SQS::Queue",
    },
    "some-nameQueue": Object {
      "Properties": Object {
        "QueueName": "some prefixsome-nameQueue",
        "RedrivePolicy": Object {
          "deadLetterTargetArn": Object {
            "Fn::GetAtt": Array [
              "some-nameDeadLetterQueue",
              "Arn",
            ],
          },
          "maxReceiveCount": 4,
        },
        "VisibilityTimeout": 4321,
      },
      "Type": "AWS::SQS::Queue",
    },
  },
}
`;
